### PR TITLE
Add GitHub Actions Workflows

### PR DESCRIPTION
**Pull request by ajschmidt8**
## Description

This PR adds GitHub Actions workflows to `cudf`.

### Task list

Coverage required for this PR:
- [x] C++ tests
- [x] Python tests for cudf (first success: c3bdbce)
- [x] Python tests for custreamz, cudf_kafka
- [x] Java tests
- [x] Check dask configuration
- [x] `strings_udf`
- [x] Benchmarks
- [x] Notebooks
- [x] Codecov
- [x] Enable style checks script (`ci/check_style.sh`)
- [x] Consolidate [existing checks job](https://github.com/rapidsai/cudf/blob/c37d33c86f86bac13cb78b1e4643b05be81115df/.github/workflows/dependency-files.yml#L7-L12) into the `pr.yaml` workflow in this PR.

Future PRs:
- [ ] Build Metrics Report
- [ ] `libidentify_stream_usage`
- [ ] JUnit output annotations? https://github.com/marketplace/actions/junit-report-action
- [ ] Reduce exit code handling verbosity (use global variable with `trap` function)
- [ ] Setup `rapids-dependency-file-generator` pre-commit hook (also setup for others repos: `rmm`, `rapids-cmake`, `cuspatial`, etc.)
- [ ] Docs builds

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

**Comment by ajschmidt8** [comment]
Adding a note so that we don't forget to include these items in this PR:

- [ ] update recipes to use `RAPIDS_CUDA_VERSION` from our CI images
  - https://github.com/rapidsai/rmm/blob/637d5fd37643c4d783da1ed1de62ba2102dd290e/conda/recipes/rmm/meta.yaml#L5
- [ ] add this temporary workaround to the Jenkins scripts to ensure the above change doesn't break Jenkins while it's still in use
  - https://github.com/rapidsai/rmm/pull/1119/commits/9b410fd00d2a65c68a862a1ba7b229c38dcdb3ca
- [ ] remove any extraneous env variables that since been added to our CI images
  - https://github.com/rapidsai/rmm/pull/1155

**Comment by bdice** [comment]
### Task list

Required for this PR:
- [x] C++ tests
- [ ] Python tests for cudf
- [ ] Python tests for custreamz, cudf_kafka
- [ ] Check dask configuration
- [ ] `strings_udf`
- [ ] Codecov

Future PRs:
- [ ] Build Metrics Report
- [ ] `libidentify_stream_usage`

**Comment by ajschmidt8** [comment]
Another thing we need to think about...

For the Python tests jobs, the logs are 100,000+ lines and over 17MB in size. This makes them unviewable in the default GH log section. Instead, you must click on `View raw logs` from the menu to see the entire log contents. I think we can cut this down significantly by removing the `-v` flag from our `pytest` invocation [here](https://github.com/ajschmidt8/cudf/blob/b5ec991475db6e2dbe0f97220b482c683d1bdcb4/ci/test_python.sh#L44). I will add this change to my next commit.

Log link for screenshot below: https://github.com/rapidsai/cudf/actions/runs/3442323626/jobs/5743299089

image

File: conda/environments/all_cuda-115_arch-x86_64.yaml [hunk 1]
```diff
@@ -58,7 +61,7 @@ dependencies:
 - python-confluent-kafka=1.7.0
 - python-snappy>=0.6.0
 - python>=3.8,<3.10
-- pytorch<1.12.0
+- pytorch
```
Comment (bdice) [inline conda/environments/all_cuda-115_arch-x86_64.yaml:64]:
```suggestion
- pytorch<1.12.0
```

File: dependencies.yaml [hunk 2]
```diff
@@ -195,13 +246,15 @@ dependencies:
...
           - moto>=4.0.8
+          - pyorc
           - pytest
           - pytest-benchmark
           - pytest-cases
           - pytest-cov
           - pytest-xdist
           - python-snappy>=0.6.0
-          - pytorch<1.12.0  # We should check and fix all "<=" pinnings
+          - pytorch
```
Comment (bdice) [inline dependencies.yaml:256]:
```suggestion
          - pytorch<1.12.0
```

**Comment by bdice** [comment]
@ajschmidt8 and I found that the Ubuntu images were missing `tzdata`, which meant that there was no `/etc/localtime`, which caused hard crashes when using the pyarrow ORC reader. This PR will add `tzdata` to all our images. https://github.com/rapidsai/mambaforge-cuda/pull/16

(Note that the CentOS 7 job had succeeded above at least once, because it comes with `tzdata` pre-installed.)

**Comment by vyasr** [comment]
@ajschmidt8 Regarding shrinking test output, you can also run `ctest` rather than looping over each test executable the way that we've done in the past [here](https://github.com/rapidsai/cudf/pull/12002/files#diff-14f652a4930c25373ff34820e67818fd5201fad17e052cd0ac205e50c93945b1R51). Using `ctest` will automatically group the output and will also be quieter (and that can be changed at any time by passing the -V verbosity flag).

File: dependencies.yaml [hunk 3]
```diff
@@ -170,19 +231,29 @@ dependencies:
...
+              - cupy-cuda11x -f https://pip.cupy.dev/aarch64 # TODO: Verify that this works.
+  test_java:
+    common:
+      - output_types: conda
+        packages:
+          # This list is copied from ci/gpu/java.sh and needs to be filtered
+          - dask-cuda=22.12.*
+          - maven
+          - openjdk=8.*
+          - rapids-build-env=22.12.*
```
Comment (ajschmidt8) [inline dependencies.yaml:254]:
```suggestion
```

`rapids-build-env` can 100% be removed. that package installs several GBs of miscellaneous dependencies.

File: conda/recipes/cudf/meta.yaml [hunk 4]
```diff
@@ -17,10 +17,22 @@ build:
...
     - PARALLEL_LEVEL
-  # libcudf's run_exports pinning is looser than we would like
+    - VERSION_SUFFIX
+    - CMAKE_GENERATOR
+    - CMAKE_C_COMPILER_LAUNCHER
+    - CMAKE_CXX_COMPILER_LAUNCHER
+    - CMAKE_CUDA_COMPILER_LAUNCHER
+    # TODO: remove "gha" suffix before PR is merged
+    - SCCACHE_S3_KEY_PREFIX=libcudf-aarch64-gha # [aarch64]
+    - SCCACHE_S3_KEY_PREFIX=libcudf-linux64-gha # [linux64]
```
Comment (bdice) [inline conda/recipes/cudf/meta.yaml:28]: Leave this suggestion open until we're ready to merge, then apply it.
```suggestion
    - SCCACHE_S3_KEY_PREFIX=libcudf-aarch64 # [aarch64]
    - SCCACHE_S3_KEY_PREFIX=libcudf-linux64 # [linux64]
```

File: conda/recipes/cudf_kafka/meta.yaml [hunk 5]
```diff
@@ -19,12 +19,25 @@ build:
   script_env:
     - PARALLEL_LEVEL
     - VERSION_SUFFIX
+    - CMAKE_GENERATOR
+    - CMAKE_C_COMPILER_LAUNCHER
+    - CMAKE_CXX_COMPILER_LAUNCHER
+    - CMAKE_CUDA_COMPILER_LAUNCHER
+    # TODO: remove "gha" suffix before PR is merged
+    - SCCACHE_S3_KEY_PREFIX=libcudf-aarch64-gha # [aarch64]
+    - SCCACHE_S3_KEY_PREFIX=libcudf-linux64-gha # [linux64]
```
Comment (bdice) [inline conda/recipes/cudf_kafka/meta.yaml:28]: Leave this suggestion open until we're ready to merge, then apply it.
```suggestion
    - SCCACHE_S3_KEY_PREFIX=libcudf-aarch64 # [aarch64]
    - SCCACHE_S3_KEY_PREFIX=libcudf-linux64 # [linux64]
```

File: conda/recipes/custreamz/meta.yaml [hunk 6]
```diff
@@ -17,8 +17,20 @@ build:
...
-    - VERSION_SUFFIX
     - PARALLEL_LEVEL
+    - VERSION_SUFFIX
+    - CMAKE_GENERATOR
+    - CMAKE_C_COMPILER_LAUNCHER
+    - CMAKE_CXX_COMPILER_LAUNCHER
+    - CMAKE_CUDA_COMPILER_LAUNCHER
+    # TODO: remove "gha" suffix before PR is merged
+    - SCCACHE_S3_KEY_PREFIX=libcudf-aarch64-gha # [aarch64]
+    - SCCACHE_S3_KEY_PREFIX=libcudf-linux64-gha # [linux64]
```
Comment (bdice) [inline conda/recipes/custreamz/meta.yaml:28]: Leave this suggestion open until we're ready to merge, then apply it.
```suggestion
    - SCCACHE_S3_KEY_PREFIX=libcudf-aarch64 # [aarch64]
    - SCCACHE_S3_KEY_PREFIX=libcudf-linux64 # [linux64]
```

File: conda/recipes/dask-cudf/meta.yaml [hunk 7]
```diff
@@ -17,8 +17,20 @@ build:
...
-    - VERSION_SUFFIX
     - PARALLEL_LEVEL
+    - VERSION_SUFFIX
+    - CMAKE_GENERATOR
+    - CMAKE_C_COMPILER_LAUNCHER
+    - CMAKE_CXX_COMPILER_LAUNCHER
+    - CMAKE_CUDA_COMPILER_LAUNCHER
+    # TODO: remove "gha" suffix before PR is merged
+    - SCCACHE_S3_KEY_PREFIX=libcudf-aarch64-gha # [aarch64]
+    - SCCACHE_S3_KEY_PREFIX=libcudf-linux64-gha # [linux64]
```
Comment (bdice) [inline conda/recipes/dask-cudf/meta.yaml:28]: Leave this suggestion open until we're ready to merge, then apply it.
```suggestion
    - SCCACHE_S3_KEY_PREFIX=libcudf-aarch64 # [aarch64]
    - SCCACHE_S3_KEY_PREFIX=libcudf-linux64 # [linux64]
```

File: conda/recipes/libcudf/meta.yaml [hunk 8]
```diff
@@ -19,22 +19,29 @@ build:
...
     - CMAKE_CXX_COMPILER_LAUNCHER
     - CMAKE_CUDA_COMPILER_LAUNCHER
-    - SCCACHE_S3_KEY_PREFIX=libcudf-aarch64 # [aarch64]
-    - SCCACHE_S3_KEY_PREFIX=libcudf-linux64 # [linux64]
-    - SCCACHE_BUCKET=rapids-sccache
-    - SCCACHE_REGION=us-west-2
-    - SCCACHE_IDLE_TIMEOUT=32768
+    # TODO: remove "gha" suffix before PR is merged
+    - SCCACHE_S3_KEY_PREFIX=libcudf-aarch64-gha # [aarch64]
+    - SCCACHE_S3_KEY_PREFIX=libcudf-linux64-gha # [linux64]
```
Comment (bdice) [inline conda/recipes/libcudf/meta.yaml:24]: Leave this suggestion open until we're ready to merge, then apply it.
```suggestion
    - SCCACHE_S3_KEY_PREFIX=libcudf-aarch64 # [aarch64]
    - SCCACHE_S3_KEY_PREFIX=libcudf-linux64 # [linux64]
```

File: conda/recipes/strings_udf/meta.yaml [hunk 9]
```diff
@@ -17,10 +17,22 @@ build:
...
     - PARALLEL_LEVEL
-  # libcudf's run_exports pinning is looser than we would like
+    - VERSION_SUFFIX
+    - CMAKE_GENERATOR
+    - CMAKE_C_COMPILER_LAUNCHER
+    - CMAKE_CXX_COMPILER_LAUNCHER
+    - CMAKE_CUDA_COMPILER_LAUNCHER
+    # TODO: remove "gha" suffix before PR is merged
+    - SCCACHE_S3_KEY_PREFIX=libcudf-aarch64-gha # [aarch64]
+    - SCCACHE_S3_KEY_PREFIX=libcudf-linux64-gha # [linux64]
```
Comment (bdice) [inline conda/recipes/strings_udf/meta.yaml:28]: Leave this suggestion open until we're ready to merge, then apply it.
```suggestion
    - SCCACHE_S3_KEY_PREFIX=libcudf-aarch64 # [aarch64]
    - SCCACHE_S3_KEY_PREFIX=libcudf-linux64 # [linux64]
```

**Review by galipremsagar — APPROVED** [review]
Dask pinnings looks good to me 👍

File: conda/recipes/cudf/meta.yaml (hunk 4 above)
Comment (ajschmidt8) [reply conda/recipes/cudf/meta.yaml]: resolved in 85b1ee1

File: conda/recipes/cudf_kafka/meta.yaml (hunk 5 above)
Comment (ajschmidt8) [reply conda/recipes/cudf_kafka/meta.yaml]: resolved in 85b1ee1

File: conda/recipes/custreamz/meta.yaml (hunk 6 above)
Comment (ajschmidt8) [reply conda/recipes/custreamz/meta.yaml]: resolved in 85b1ee1

File: conda/recipes/libcudf/meta.yaml (hunk 8 above)
Comment (ajschmidt8) [reply conda/recipes/libcudf/meta.yaml]: resolved in 85b1ee1

File: conda/recipes/strings_udf/meta.yaml (hunk 9 above)
Comment (ajschmidt8) [reply conda/recipes/strings_udf/meta.yaml]: resolved in 85b1ee1

File: conda/recipes/dask-cudf/meta.yaml (hunk 7 above)
Comment (ajschmidt8) [reply conda/recipes/dask-cudf/meta.yaml]: resolved in 85b1ee1

**Comment by ajschmidt8** [comment]
rerun tests

**Review by ttnghia — APPROVED** [review]

File: ci/test_cpp.sh [hunk 10]
```diff
@@ -0,0 +1,101 @@
...
+
+rapids-mamba-retry env create --force -f env.yaml -n test
+
+# Temporarily allow unbound variables for conda activation.
+set +u
+conda activate test
+set -u
+
+CPP_CHANNEL=$(rapids-download-conda-from-s3 cpp)
+RAPIDS_TESTS_DIR=${RAPIDS_TESTS_DIR:-"${PWD}/test-results"}/
```
Comment (bdice) [inline ci/test_cpp.sh:22]: I don’t like adding the slash here. This line should only handle the default value logic. I’d put the slash in the original environment variable definition and make this change:
```suggestion
RAPIDS_TESTS_DIR=${RAPIDS_TESTS_DIR:-"${PWD}/test-results/"}
```

Comment (bdice) [reply ci/test_cpp.sh:22]: Or maybe add the slash in the places where it’s needed below? Then we can avoid the missing slash in `"${RAPIDS_TESTS_DIR}${test_name}.cs.log"`.

File: conda/recipes/cudf/meta.yaml [hunk 11]
```diff
@@ -17,10 +17,22 @@ build:
...
   script_env:
-    - VERSION_SUFFIX
     - PARALLEL_LEVEL
-  # libcudf's run_exports pinning is looser than we would like
+    - VERSION_SUFFIX
+    - CMAKE_GENERATOR
+    - CMAKE_C_COMPILER_LAUNCHER
+    - CMAKE_CXX_COMPILER_LAUNCHER
+    - CMAKE_CUDA_COMPILER_LAUNCHER
+    # TODO: remove "gha" suffix before PR is merged
```
Comment (bdice) [inline conda/recipes/cudf/meta.yaml:26]:
```suggestion
```

File: conda/recipes/cudf_kafka/meta.yaml [hunk 12]
```diff
@@ -19,12 +19,25 @@ build:
   script_env:
     - PARALLEL_LEVEL
     - VERSION_SUFFIX
+    - CMAKE_GENERATOR
+    - CMAKE_C_COMPILER_LAUNCHER
+    - CMAKE_CXX_COMPILER_LAUNCHER
+    - CMAKE_CUDA_COMPILER_LAUNCHER
+    # TODO: remove "gha" suffix before PR is merged
```
Comment (bdice) [inline conda/recipes/cudf_kafka/meta.yaml:26]:
```suggestion
```

File: conda/recipes/custreamz/meta.yaml [hunk 13]
```diff
@@ -17,8 +17,20 @@ build:
...
   string: py{{ py_version_numeric }}_{{ GIT_DESCRIBE_HASH }}_{{ GIT_DESCRIBE_NUMBER }}
   script_env:
-    - VERSION_SUFFIX
     - PARALLEL_LEVEL
+    - VERSION_SUFFIX
+    - CMAKE_GENERATOR
+    - CMAKE_C_COMPILER_LAUNCHER
+    - CMAKE_CXX_COMPILER_LAUNCHER
+    - CMAKE_CUDA_COMPILER_LAUNCHER
+    # TODO: remove "gha" suffix before PR is merged
```
Comment (bdice) [inline conda/recipes/custreamz/meta.yaml:26]:
```suggestion
```

File: conda/recipes/dask-cudf/meta.yaml [hunk 14]
```diff
@@ -17,8 +17,20 @@ build:
...
   string: cuda_{{ cuda_major }}_py{{ py_version }}_{{ GIT_DESCRIBE_HASH }}_{{ GIT_DESCRIBE_NUMBER }}
   script_env:
-    - VERSION_SUFFIX
     - PARALLEL_LEVEL
+    - VERSION_SUFFIX
+    - CMAKE_GENERATOR
+    - CMAKE_C_COMPILER_LAUNCHER
+    - CMAKE_CXX_COMPILER_LAUNCHER
+    - CMAKE_CUDA_COMPILER_LAUNCHER
+    # TODO: remove "gha" suffix before PR is merged
```
Comment (bdice) [inline conda/recipes/dask-cudf/meta.yaml:26]:
```suggestion
```

File: conda/recipes/libcudf/meta.yaml [hunk 15]
```diff
@@ -19,22 +19,29 @@ build:
     - CMAKE_C_COMPILER_LAUNCHER
     - CMAKE_CXX_COMPILER_LAUNCHER
     - CMAKE_CUDA_COMPILER_LAUNCHER
+    # TODO: remove "gha" suffix before PR is merged
```
Comment (bdice) [inline conda/recipes/libcudf/meta.yaml:22]:
```suggestion
```

File: conda/recipes/strings_udf/meta.yaml [hunk 16]
```diff
@@ -17,10 +17,22 @@ build:
...
   script_env:
-    - VERSION_SUFFIX
     - PARALLEL_LEVEL
-  # libcudf's run_exports pinning is looser than we would like
+    - VERSION_SUFFIX
+    - CMAKE_GENERATOR
+    - CMAKE_C_COMPILER_LAUNCHER
+    - CMAKE_CXX_COMPILER_LAUNCHER
+    - CMAKE_CUDA_COMPILER_LAUNCHER
+    # TODO: remove "gha" suffix before PR is merged
```
Comment (bdice) [inline conda/recipes/strings_udf/meta.yaml:26]:
```suggestion
```

**Review by bdice — COMMENTED** [review]
A few minor edits.

**Review by Ethyling — APPROVED** [review]

File: conda/recipes/cudf/meta.yaml (hunk 11 above)
Comment (ajschmidt8) [reply conda/recipes/cudf/meta.yaml:26]: whoops. :facepalm:. i'll go back and remove these after the current builds are complete and the cache is re-populated.

File: ci/test_cpp.sh (hunk 10 above)
Comment (ajschmidt8) [reply ci/test_cpp.sh:22]: I was torn on this...

> Or maybe add the slash in the places where it’s needed below?

The trailing slash is actually needed everywhere. Without the trailing slash, the C++ tests throw this error https://github.com/rapidsai/cudf/actions/runs/3586797596/jobs/6036781772#step:6:66377.

So I thought about putting the trailing slash in the shared workflow variable definition below, but then there's a discrepancy between the `RAPIDS_COVERAGE_DIR` and `RAPIDS_TESTS_DIR` variables (e.g. one has a trailing slash, but the other doesn't).

https://github.com/rapidsai/shared-action-workflows/blob/4bf344150bf50071678fe9d04774931ebb974f78/.github/workflows/conda-python-tests.yaml#L87-L88

I'm open to changing this if you want, but there are implications wherever we decide to move it.

**Review by bdice — APPROVED** [review]

File: conda/recipes/strings_udf/meta.yaml [hunk 17]
```diff
@@ -50,12 +62,12 @@ requirements:
...
     - {{ pin_compatible('cudatoolkit', max_pin='x', min_pin='x') }}
     - cachetools
-    - ptxcompiler >=0.7.0  # [linux64]  # CUDA enhanced compatibility. See https://github.com/rapidsai/ptxcompiler
-test:                                   # [linux64]
-  requires:                             # [linux64]
-    - cudatoolkit {{ cuda_version }}.*  # [linux64]
-  imports:                              # [linux64]
-    - strings_udf                       # [linux64]
+    - ptxcompiler >=0.7.0  # CUDA enhanced compatibility. See https://github.com/rapidsai/ptxcompiler
+test:
```
Comment (bdice) [inline conda/recipes/strings_udf/meta.yaml:66]: This is failing on ARM + gpuCI Jenkins. We probably can't remove the `# [linux64]` until Jenkins is disabled. This appears to work fine on GitHub Actions for whatever reason.

**Pull request merged**